### PR TITLE
[crypto] Unify Narwhal and Sui Crypto

### PR DESCRIPTION
**Pull request by punwai**
**This PR relies on**: https://github.com/MystenLabs/narwhal/pull/460/files

- All implementations of cryptographic primitives are moved over to the Narwhal repository (except AccountSignatures/Signature for now). Authority Signatures are now completely generic.
- Added 'batch verification' on another PR in narwhal that will allow us to plug any batched signature into Sui
- More generic and cleaner version of #2877

Next to dos:
- Change PublicKeyBytes to an Associated Type? This will massively help me out when implementing other things

### CODE REVIEW

File: crates/sui-types/src/crypto.rs
Comment (punwai) [inline crates/sui-types/src/crypto.rs:63]: Should 100% be a sui feature

File: crates/sui-types/src/crypto.rs
Comment (punwai) [inline crates/sui-types/src/crypto.rs:71]: Keep in Sui

File: crates/sui-types/src/crypto.rs
Comment (punwai) [inline crates/sui-types/src/crypto.rs:76]: transfer to Narwhal

File: crates/sui-types/src/crypto.rs
Comment (punwai) [inline crates/sui-types/src/crypto.rs:90]: Keep in Sui

File: crates/sui-types/src/crypto.rs
Comment (punwai) [inline crates/sui-types/src/crypto.rs:15]: Will add this back to the repo, but through Narwhal itself.

File: crates/sui-types/src/crypto.rs
Comment (punwai) [inline crates/sui-types/src/crypto.rs:33]: Add this into Narwhal as well

File: crates/sui-types/src/crypto.rs
Comment (punwai) [inline crates/sui-types/src/crypto.rs:45]: Shift this to the Narwhal repo

File: crates/sui-types/src/crypto.rs
Comment (punwai) [inline crates/sui-types/src/crypto.rs:51]: Shift to Narwhal Repo

**Comment by punwai** [comment]
@kchalkias @huitseeker Would love a first look through this!

File: crates/sui-benchmark/src/benchmark/bench_types.rs
Comment (huitseeker) [inline crates/sui-benchmark/src/benchmark/bench_types.rs:94]: This should not be changed.

File: crates/sui-types/src/base_types.rs
Comment (huitseeker) [inline crates/sui-types/src/base_types.rs:188]: As discussed, we'd like to leave this to a distinct PR.

**Review by huitseeker — COMMENTED** [review]
Promising ! 🚀

File: crates/sui-benchmark/src/benchmark/validator_preparer.rs
Comment (punwai) [reply crates/sui-benchmark/src/benchmark/validator_preparer.rs:25]: NarwhalKeypair is the trait

File: crates/sui-types/src/base_types.rs
Comment (punwai) [reply crates/sui-types/src/base_types.rs]: Removed and done.

File: crates/sui-benchmark/Cargo.toml
Comment (kchalkias) [inline crates/sui-benchmark/Cargo.toml:35]: Are we expecting to make updates to Narwhal first and then pin the official crate?

File: crates/sui-types/src/crypto.rs
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:31]: Do we want to merge this first then allow crypto agility? I.e., eventually the keypair can be of any supported type (ECDSA, BLS, EdDSA etc)

Comment (kchalkias) [reply crates/sui-types/src/crypto.rs:31]: It also doesn't harm on the future to have two key types (one for users and another for validators), mostly because we'll allow different schemes per type (ie we don't expect validators to ever support ECDSA)

File: crates/sui-types/src/crypto.rs
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:33]: The main reason for this existence of this memoizer pattern is to to avoid converting to bytes each time.

File: crates/sui-types/src/crypto.rs
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:145]: is this removed because narwhal supports this already?

File: crates/sui-types/src/crypto.rs
Comment (kchalkias) [inline crates/sui-types/src/crypto.rs:201]: Note that in crypto agility we cannot have a global fixed length (each scheme has (might have) different size)

File: crates/sui-types/src/unit_tests/base_types_tests.rs
Comment (kchalkias) [inline crates/sui-types/src/unit_tests/base_types_tests.rs:370]: As we can see, if we made a non compatible change (and we know this was on purpose) we have to update this test Vs removing it.

File: crates/sui-benchmark/Cargo.toml
Comment (punwai) [reply crates/sui-benchmark/Cargo.toml:35]: Yes, this is until the PR gets merged.

File: crates/sui-types/src/crypto.rs
Comment (punwai) [reply crates/sui-types/src/crypto.rs:31]: I have another PR that will address this!

File: crates/sui-types/src/crypto.rs
Comment (punwai) [reply crates/sui-types/src/crypto.rs:145]: It is removed because Narwhal currently does not support it, I can also add this to Narwhal and uncomment this for the PR.

File: crates/sui-types/src/crypto.rs
Comment (punwai) [reply crates/sui-types/src/crypto.rs:33]: I will send in a PR to Narwhal that has these changes!

File: crates/sui-types/src/crypto.rs
Comment (punwai) [reply crates/sui-types/src/crypto.rs]: I added it back in.

File: crates/sui-types/src/crypto.rs
Comment (huitseeker) [reply crates/sui-types/src/crypto.rs:33]: There is no need for this memoizer pattern on Ed25519. 

Here's a dalek `PublicKey`: https://github.com/dalek-cryptography/ed25519-dalek/blob/main/src/public.rs#L40
the `CompressedEdwards`: https://github.com/dalek-cryptography/curve25519-dalek/blob/main/src/edwards.rs#L164

The fact that a `Foo` implements `AsRef<[u8]>` exactly represents that you can get a zero-copy reference to its byte representation.

`dalek_key_pair.public.as_ref::<[u8]>()` involves exactly zero byte conversions.

Comment (punwai) [reply crates/sui-types/src/crypto.rs:33]: This abstraction is done in the narwhal library :)

File: crates/sui-types/src/crypto.rs
Comment (huitseeker) [reply crates/sui-types/src/crypto.rs:49]: I see. I think it's possible to make this an extension trait of `Authenticator`.

File: crates/sui-types/src/crypto.rs
Comment (huitseeker) [inline crates/sui-types/src/crypto.rs:93]:
```rust
        let signature_bytes: AccountSignature =
            <Self as signature::Signer<AccountSignature>>::try_sign(self, msg)?;
        let public_key_bytes: PublicKeyBytes = self.public().into();
        let mut result_bytes = [0u8; SUI_SIGNATURE_LENGTH];
        let sig_length = <KeyPair as narwhal_crypto::traits::KeyPair>::Sig::LENGTH;
        result_bytes[..sig_length].copy_from_slice(signature_bytes.as_ref());
        result_bytes[sig_length..].copy_from_slice(public_key_bytes.as_ref());
        Ok(Signature(result_bytes))
```

File: crates/sui-types/src/crypto.rs
Comment (huitseeker) [inline crates/sui-types/src/crypto.rs:227]:
```suggestion
       let public_key = PublicKey::from_bytes(public_key_bytes.as_ref()).map_err(|err| {
            SuiError::InvalidSignature {
                error: err.to_string(),
            }
        })?;
```

**Review by huitseeker — CHANGES_REQUESTED** [review]
Thanks! This is a huge amount of necessary work, I'm very grateful for the pretty advanced stage this is at.

I've left a few changes here: https://github.com/MystenLabs/sui/pull/3293

I'm pretty sure by now that `KeypairTraits` is not needed and should be removed from https://github.com/MystenLabs/narwhal/pull/507. There are some traits that are indeed needed (e.g. `Send` and `Sync`) but those should be added to the `KeyPair` trait, which should be imported to get the affordance of the `KeyPair` functions.

File: crates/sui-core/src/generate_format.rs
Comment (punwai) [inline crates/sui-core/src/generate_format.rs:78]: Getting an error here that I do not quite understand `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Custom("Failed to deserialize value: \"invalid length 0, expected an byte array of size 32\"")', crates/sui-core/src/generate_format.rs:104:35`

File: crates/sui-types/src/crypto.rs
Comment (punwai) [reply crates/sui-types/src/crypto.rs:121]: I don't think this is longer possible/useful because we removed the generics.

**Review by kchalkias — APPROVED** [review]
Amazing! good to go

**Pull request merged**
